### PR TITLE
Puppeteer code samples

### PR DESCRIPTION
**Pull request by sourav-kundu**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+node_modules
```
Comment (Archish27) [inline .gitignore:1]: EOL

File: README.md [hunk 2]
```diff
@@ -1,2 +1,78 @@
...
+
+While your Puppeteer session runs on BrowserStack, we generate a unique ID for the session, build and also generate URLs for the various types of logs which you can use to build your own reporting or for any other purpose that you might like.
+
+A sample script with the use of the API to fetch all the relevant session details is shown in the [sample_session_details_API.js](./sample_session_details_API.js). 
+
+You can see the [documentation for marking test status using REST API](https://www.browserstack.com/docs/automate/api-reference/selenium/session#set-test-status) using the session ID for the session, any time even after the session has completed its execution.
+
+## Facing issues?
+
+If you are facing any issue with any of the above or any other issue in trying to run your Playwright tests on BrowserStack, you can reach out to me directly at `sourav.k@browserstack.com` and I will be happy to debug your issues or at the least ensure that your issue becomes our top priority to resolve.
```
Comment (Archish27) [inline README.md:78]: Is this purposely we are adding your email id or is a support email id is required here?

File: codeceptjs-example/codecept.conf.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      enabled: true
+    },
+    tryTo: {
+      enabled: true
+    },
+    screenshotOnFail: {
+      enabled: true
+    }
+  }
+}
```
Comment (Archish27) [inline codeceptjs-example/codecept.conf.js:46]: EOL

File: codeceptjs-example/jsconfig.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
+  "compilerOptions": {
+    "allowJs": true
+  }
+}
```
Comment (Archish27) [inline codeceptjs-example/jsconfig.json:5]: EOL

File: google_search.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    console.log(title);
+    try {
+        expect(title).to.equal("BrowserStack - Google Search", 'Expected page title is incorrect!');
+        // following line of code is responsible for marking the status of the test on BrowserStack as 'passed'. You can use this code in your after hook after each test
+        await page.evaluate(_ => {}, `browserstack_executor: ${JSON.stringify({action: 'setSessionStatus',arguments: {status: 'passed',reason: 'Title matched'}})}`);
+    } catch {
+        await page.evaluate(_ => {}, `browserstack_executor: ${JSON.stringify({action: 'setSessionStatus',arguments: {status: 'failed',reason: 'Title did not match'}})}`);
+    }
+    await browser.close();
+})();
```
Comment (Archish27) [inline google_search.js:40]: EOL

File: google_search_with_proxy.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    console.log(title);
+    try {
+        expect(title).to.equal("BrowserStack - Google Search", 'Expected page title is incorrect!');
+        // following line of code is responsible for marking the status of the test on BrowserStack as 'passed'. You can use this code in your after hook after each test
+        await page.evaluate(_ => {}, `browserstack_executor: ${JSON.stringify({action: 'setSessionStatus',arguments: {status: 'passed',reason: 'Title matched'}})}`);
+    } catch {
+        await page.evaluate(_ => {}, `browserstack_executor: ${JSON.stringify({action: 'setSessionStatus',arguments: {status: 'failed',reason: 'Title did not match'}})}`);
+    }
+    await browser.close();
+})();
```
Comment (Archish27) [inline google_search_with_proxy.js:47]: EOL

File: local_test.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        expect(title).to.equal("BrowserStack Local", 'Expected page title is incorrect!');
+        // following line of code is responsible for marking the status of the test on BrowserStack as 'passed'. You can use this code in your after hook after each test
+        await page.evaluate(_ => {}, `browserstack_executor: ${JSON.stringify({action: 'setSessionStatus',arguments: {status: 'passed',reason: 'Local is up and running'}})}`);
+    } catch {
+        await page.evaluate(_ => {}, `browserstack_executor: ${JSON.stringify({action: 'setSessionStatus',arguments: {status: 'failed',reason: 'BrowserStack Local binary is not running'}})}`);
+    }
+    await browser.close();
+})();
+
+//  After the end of the script, ensure to stop the BrowserStackLocal binary that had been started before invoking this script.
```
Comment (Archish27) [inline local_test.js:40]: EOL

File: parallel_test.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+      } catch {
+        await page.evaluate(_ => {}, `browserstack_executor: ${JSON.stringify({action: 'setSessionStatus',arguments: {status: 'failed',reason: 'Title did not match'}})}`);
+      }
+    await browser.close();
+};
+
+//  The following capabilities array contains the list of os/browser environments where you want to run your tests. You can choose to alter this list according to your needs
+const capabilities = [
+{
+  	'browser': 'chrome',
```
Comment (Archish27) [inline parallel_test.js:34]: Indentation fix required here

File: parallel_test.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    'browser_version': 'latest',
+    'os': 'osx',
+    'os_version': 'catalina',
+    'name': 'Chrome latest on Catalina',
+    'build': 'puppeteer-build-2'
+},
+{
+  	'browser': 'firefox',
```
Comment (Archish27) [inline parallel_test.js:42]: Same here

File: parallel_test.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  'os': 'Windows',
+  'os_version': '10',
+  'name': 'Edge latest on Win10',
+  'build': 'puppeteer-build-2'
+}]
+
+//  The following code loops through the capabilities array defined above and runs your code against each environment that you have specified
+capabilities.forEach(async (cap) => {
+  await main(cap);
+});
```
Comment (Archish27) [inline parallel_test.js:85]: EOL

File: puppeteer-jest/google.test.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    await element.type('BrowserStack');
+    await element.press('Enter');
+    try {
+      expect(await page.title()).toBe('BrowserStack - Google Search');
+      await page.evaluate(_ => {}, `browserstack_executor: ${JSON.stringify({action: 'setSessionStatus',arguments: {status: 'passed',reason: 'Test assertion passed'}})}`);
+    } catch {
+      await page.evaluate(_ => {}, `browserstack_executor: ${JSON.stringify({action: 'setSessionStatus',arguments: {status: 'failed',reason: 'Test assertion failed'}})}`);
+    }
+  })
+})
```
Comment (Archish27) [inline puppeteer-jest/google.test.js:21]: EOL

File: puppeteer-jest/jest-puppeteer.config.js [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    'build': 'puppeteer-jest-build-2',
+    'browserstack.username': process.env.BROWSERSTACK_USERNAME || 'YOUR_USERNAME',
+    'browserstack.accessKey': process.env.BROWSERSTACK_ACCESS_KEY || 'YOUR_ACCESS_KEY'
+};
+
+module.exports = {
+  connect: {
+    browserWSEndpoint: `ws://cdp.browserstack.com?caps=${encodeURIComponent(JSON.stringify(caps_chrome))}`,
+  }
+}
```
Comment (Archish27) [inline puppeteer-jest/jest-puppeteer.config.js:38]: EOL

File: puppeteer-jest/jest-puppeteer.config.js [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    'os_version': 'big sur',
+    'name': 'Puppeteer-jest test on Edge',
+    'build': 'puppeteer-jest-build-2',
+    'browserstack.username': process.env.BROWSERSTACK_USERNAME || 'YOUR_USERNAME',
+    'browserstack.accessKey': process.env.BROWSERSTACK_ACCESS_KEY || 'YOUR_ACCESS_KEY'
+};
+
+module.exports = {
+  connect: {
+    browserWSEndpoint: `ws://cdp.browserstack.com?caps=${encodeURIComponent(JSON.stringify(caps_chrome))}`,
```
Comment (Archish27) [inline puppeteer-jest/jest-puppeteer.config.js:36]:
```suggestion
    browserWSEndpoint: `wss://cdp.browserstack.com?caps=${encodeURIComponent(JSON.stringify(caps_chrome))}`,
```

File: sample_session_details_API.js [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        'os': 'os x',
+        'os_version': 'big sur',
+        'build': 'puppeteer-build-1',
+        'name': 'My first Puppeteer test',
+        'browserstack.username': process.env.BROWSERSTACK_USERNAME || 'YOUR_USERNAME',
+        'browserstack.accessKey': process.env.BROWSERSTACK_ACCESS_KEY || 'YOUR_ACCESS_KEY'
+    };
+    const browser = await puppeteer.connect({
+    browserWSEndpoint:
+    `ws://cdp.browserstack.com?caps=${encodeURIComponent(JSON.stringify(caps))}`,
```
Comment (Archish27) [inline sample_session_details_API.js:18]:
```suggestion
    `wss://cdp.browserstack.com?caps=${encodeURIComponent(JSON.stringify(caps))}`,
```

File: sample_session_details_API.js [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    *  The following part of the code uses the getSessionDetails API to get all the relevant details about the running playwright session.
+    *  You can use all these details after your test has completed, to fetch logs or use any other Automate REST APIs
+    */
+    const resp = await JSON.parse(await page.evaluate(_ => {}, `browserstack_executor: ${JSON.stringify({action: 'getSessionDetails'})}`));
+    const jsonObj = await JSON.parse(resp);
+    console.log(jsonObj.hashed_id);  // This gives the session ID of the running session
+    console.log(jsonObj);  // This prints the entire JSON response. You can use any/all of the response attributes the way you like.
+    
+    await browser.close();
+})();
```
Comment (Archish27) [inline sample_session_details_API.js:49]: EOL

File: google_search.js [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+        'os': 'os x',
+        'os_version': 'big sur',
+        'build': 'puppeteer-build-1',
+        'name': 'My first Puppeteer test',
+        'browserstack.username': process.env.BROWSERSTACK_USERNAME || 'YOUR_USERNAME',
+        'browserstack.accessKey': process.env.BROWSERSTACK_ACCESS_KEY || 'YOUR_ACCESS_KEY'
+    };
+    const browser = await puppeteer.connect({
+    browserWSEndpoint:
+    `ws://cdp.browserstack.com?caps=${encodeURIComponent(JSON.stringify(caps))}`,
```
Comment (Archish27) [inline google_search.js:18]:
```suggestion
    `wss://cdp.browserstack.com?caps=${encodeURIComponent(JSON.stringify(caps))}`,
```

File: google_search_with_proxy.js [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+        'os': 'os x',
+        'os_version': 'big sur',
+        'build': 'puppeteer-build-1',
+        'name': 'My first Puppeteer test',
+        'browserstack.username': process.env.BROWSERSTACK_USERNAME || 'YOUR_USERNAME',
+        'browserstack.accessKey': process.env.BROWSERSTACK_ACCESS_KEY || 'YOUR_ACCESS_KEY'
+    };
+    const browser = await puppeteer.connect({
+    browserWSEndpoint:
+    `ws://cdp.browserstack.com?caps=${encodeURIComponent(JSON.stringify(caps))}`,
```
Comment (Archish27) [inline google_search_with_proxy.js:25]:
```suggestion
    `wss://cdp.browserstack.com?caps=${encodeURIComponent(JSON.stringify(caps))}`,
```

File: local_test.js [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+        'os_version': 'big sur',
+        'build': 'puppeteer-build-1',
+        'name': 'My first Puppeteer test',
+        'browserstack.username': process.env.BROWSERSTACK_USERNAME || 'YOUR_USERNAME',
+        'browserstack.accessKey': process.env.BROWSERSTACK_ACCESS_KEY || 'YOUR_ACCESS_KEY',
+        'browserstack.local': 'true'  // You have to ensure that you have started the BrowserStackLocal binary in your system and you have seen '[SUCCESS] You can now access your local server(s) in our remote browser' in the terminal
+    };
+    const browser = await puppeteer.connect({
+    browserWSEndpoint:
+    `ws://cdp.browserstack.com?caps=${encodeURIComponent(JSON.stringify(caps))}`,
```
Comment (Archish27) [inline local_test.js:19]:
```suggestion
    `wss://cdp.browserstack.com?caps=${encodeURIComponent(JSON.stringify(caps))}`,
```

File: parallel_test.js [hunk 19]
```diff
@@ -0,0 +1,85 @@
+const puppeteer = require('puppeteer');
+const expect = require('chai').expect;
+const main = async (cap) => {
+    console.log("Starting test -->", cap['name'])
+    
+    cap['browserstack.username'] = process.env.BROWSERSTACK_USERNAME || 'YOUR_USERNAME';
+    cap['browserstack.accessKey'] = process.env.BROWSERSTACK_ACCESS_KEY || 'YOUR_ACCESS_KEY';
+
+    const browser = await puppeteer.connect({
+      browserWSEndpoint:`ws://cdp.browserstack.com?caps=${encodeURIComponent(JSON.stringify(cap))}`,
```
Comment (Archish27) [inline parallel_test.js:10]:
```suggestion
      browserWSEndpoint:`wss://cdp.browserstack.com?caps=${encodeURIComponent(JSON.stringify(cap))}`,
```

**Review by Archish27 — COMMENTED** [review]
Minor comments. Rest looks good 🚀

File: README.md (hunk 2 above)
Comment (sourav-kundu) [reply README.md:78]: readme was not changed...my bad...making necessary corrections

**Review by Archish27 — APPROVED** [review]
LGTM

**Review by francisf — APPROVED** [review]

**Pull request merged**